### ISSUE TITLE
kapacitor architecture/design notes

### ISSUE THREAD

**Issue by beckettsean** [influxdata/docs.influxdata.com#869]
From: https://groups.google.com/d/msgid/influxdb/736f8f2a-0820-4062-aa88-530ad1e36fd8%40googlegroups.com

> 1. is there any limit on the number of scripts kapacitor can handle ?

1. No software limit, but it will be limited by available resources. 

> 2. is it better to aim for more complex scripts or is it fine to have 2 or 3 running in parallel performance wise ?

2. Taking things to the extreme the best case is one task that consumes all the data and does all the work, as there will be less overhead etc of managing multiple tasks. That said significant effort has gone into making sure that each new task does not add too much overhead. I would recommend using tasks in a way that makes logical organizational sense first, which probably means many tasks. Then if you do run into performance issues, let's first see if we can fix them within Kapacitor, then as a last resort would I start merging tasks into more complex tasks.

> 3. are templates based scripts using less resources ? more efficient ? or is it just a ease of use tool and it make no differences below ?

3. Templates are just an ease of use tool and make no difference.

> 4. What happen when you update a script, let's say most of it did not change but I added an output, if alarms were up before the update will kapacitor remembers it ? How are scripts linked to the alert they raised and what can "break" this link ? I ask this because in most if not all cases when I update a script I want it to keep tracks of any alert raised before.

4. Currently when you update a script all state about that task is lost. Specifically for the alert this means they can be left in a triggering state :(. We are working to fix this soon as part of a new alerting system.

**Comment by stevebang** [comment]
@sanderson -- See if any of these notes/discussion need to be doc'ed or close this issue.

**Comment by sanderson** [comment]
Quick note on question 4:

Kapacitor will now remember the last level of an alert, but other state-like data, such as data buffered in a window, is still lost.
